File: src/main/templates/typescript_sdk/src/api/RecordsApi.ts
import { ApiClient } from '../client/ApiClient.js';
import { PageResult } from './PageResult.js';
import { PaginationOptions } from '../types/index.js';

/**
 * 通用Records API
 */
export class RecordsApi {
  private apiClient: ApiClient;

  constructor(apiClient: ApiClient) {
    this.apiClient = apiClient;
  }

  /**
   * 获取记录列表（分页）
   */
  async listRecords(
    datasourceName: string, 
    modelName: string, 
    options: PaginationOptions = {}
  ): Promise<PageResult<Record<string, any>>> {
    const {
      current = null,
      pageSize = null,
      filter = null,
      nestedQuery = null,
      sort = null
    } = options;

    const path = `/api/f/datasources/${this.encode(datasourceName)}/models/${this.encode(modelName)}/records`;
    const params: Record<string, any> = {};
    
    if (current !== null) params.current = current;
    if (pageSize !== null) params.pageSize = pageSize;
    if (filter !== null) params.filter = filter;
    if (nestedQuery !== null) params.nestedQuery = nestedQuery;
    if (sort !== null) params.sort = sort;

    const data = await this.apiClient.get<{ total: number; list: Record<string, any>[] }>(path, params);
    return new PageResult(data.total, data.list);
  }

  /**
   * 获取记录列表（直接返回列表）
   */
  async listRecordsAsList(
    datasourceName: string, 
    modelName: string, 
    options: PaginationOptions = {}
  ): Promise<Record<string, any>[]> {
    const pageResult = await this.listRecords(datasourceName, modelName, options);
    return pageResult.getList();
  }

  /**
   * 创建记录
   */
  async createRecord(
    datasourceName: string, 
    modelName: string, 
    body: Record<string, any>
  ): Promise<Record<string, any>> {
    const path = `/api/f/datasources/${this.encode(datasourceName)}/models/${this.encode(modelName)}/records`;
    return await this.apiClient.post<Record<string, any>>(path, body);
  }

  /**
   * 获取单个记录
   */
  async getRecord(
    datasourceName: string, 
    modelName: string, 
    id: string, 
    options: { nestedQuery?: boolean } = {}
  ): Promise<Record<string, any>> {
    const { nestedQuery = null } = options;
    const path = `/api/f/datasources/${this.encode(datasourceName)}/models/${this.encode(modelName)}/records/${this.encode(id)}`;
    const params: Record<string, any> = {};
    
    if (nestedQuery !== null) params.nestedQuery = nestedQuery;

    return await this.apiClient.get<Record<string, any>>(path, params);
  }

  /**
   * 更新记录
   */
  async updateRecord(
    datasourceName: string, 
    modelName: string, 
    id: string, 
    body: Record<string, any>
  ): Promise<Record<string, any>> {
    const path = `/api/f/datasources/${this.encode(datasourceName)}/models/${this.encode(modelName)}/records/${this.encode(id)}`;
    return await this.apiClient.put<Record<string, any>>(path, body);
  }

  /**
   * 部分更新记录
   */
  async patchRecord(
    datasourceName: string, 
    modelName: string, 
    id: string, 
    body: Record<string, any>
  ): Promise<Record<string, any>> {
    const path = `/api/f/datasources/${this.encode(datasourceName)}/models/${this.encode(modelName)}/records/${this.encode(id)}`;
    return await this.apiClient.patch<Record<string, any>>(path, body);
  }

  /**
   * 删除记录
   */
  async deleteRecord(datasourceName: string, modelName: string, id: string): Promise<void> {
    const path = `/api/f/datasources/${this.encode(datasourceName)}/models/${this.encode(modelName)}/records/${this.encode(id)}`;
    await this.apiClient.delete(path);
  }

  /**
   * URL编码
   */
  private encode(str: string): string {
    return str.replace(/\//g, '%2F');
  }
}

File: src/main/templates/java_sdk/src/main/java/${path}/api/RecordsApi.java
package ${packageName};

import com.fasterxml.jackson.core.type.TypeReference;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import ${packageName}.client.ApiClient;
import ${packageName}.client.ApiClient;


public class RecordsApi {
  private final ApiClient apiClient;

  public RecordsApi(ApiClient apiClient) {
    this.apiClient = apiClient;
  }

  // 通用方法 - 返回分页结果
  public PageResult<Map<String, Object>> listRecords(
      String datasourceName,
      String modelName,
      Integer current,
      Integer pageSize,
      String filter,
      Boolean nestedQuery,
      String sort
  ) {
    String path = String.format(
        "/api/f/datasources/%s/models/%s/records",
        encode(datasourceName), encode(modelName)
    );
    Map<String, Object> q = new HashMap<>();
    if (current != null) q.put("current", current);
    if (pageSize != null) q.put("pageSize", pageSize);
    if (filter != null) q.put("filter", filter);
    if (nestedQuery != null) q.put("nestedQuery", nestedQuery);
    if (sort != null) q.put("sort", sort);

    return apiClient.get(
        path, q, new TypeReference<PageResult<Map<String, Object>>>() {}
    );
  }

  // 通用方法 - 直接返回实体列表（不分页）
  public List<Map<String, Object>> listRecordsAsList(
      String datasourceName,
      String modelName,
      String filter,
      Boolean nestedQuery,
      String sort
  ) {
    PageResult<Map<String, Object>> pageResult = listRecords(
        datasourceName, modelName, null, null, filter, nestedQuery, sort
    );
    return pageResult.getList();
  }

  public Map<String, Object> createRecord(
      String datasourceName,
      String modelName,
      Map<String, Object> body
  ) {
    String path = String.format(
        "/api/f/datasources/%s/models/%s/records",
        encode(datasourceName), encode(modelName)
    );
    return apiClient.post(path, body, new TypeReference<Map<String, Object>>() {});
  }

  public Map<String, Object> getRecord(
      String datasourceName,
      String modelName,
      String id,
      Boolean nestedQuery
  ) {
    String path = String.format(
        "/api/f/datasources/%s/models/%s/records/%s",
        encode(datasourceName), encode(modelName), encode(id)
    );
    Map<String, Object> q = new HashMap<>();
    if (nestedQuery != null) q.put("nestedQuery", nestedQuery);
    return apiClient.get(path, q, new TypeReference<Map<String, Object>>() {});
  }

  public Map<String, Object> updateRecord(
      String datasourceName,
      String modelName,
      String id,
      Map<String, Object> body
  ) {
    String path = String.format(
        "/api/f/datasources/%s/models/%s/records/%s",
        encode(datasourceName), encode(modelName), encode(id)
    );
    return apiClient.put(path, body, new TypeReference<Map<String, Object>>() {});
  }

  public Map<String, Object> patchRecord(
      String datasourceName,
      String modelName,
      String id,
      Map<String, Object> body
  ) {
    String path = String.format(
        "/api/f/datasources/%s/models/%s/records/%s",
        encode(datasourceName), encode(modelName), encode(id)
    );
    return apiClient.patch(path, body, new TypeReference<Map<String, Object>>() {});
  }

  public void deleteRecord(
      String datasourceName,
      String modelName,
      String id
  ) {
    String path = String.format(
        "/api/f/datasources/%s/models/%s/records/%s",
        encode(datasourceName), encode(modelName), encode(id)
    );
    apiClient.delete(path);
  }

  private String encode(String s) {
    return s.replace("/", "%2F");
  }
}


